### PR TITLE
[SPARK-28158][SQL] Hive UDFs supports UDT type

### PR DESCRIPTION
**Pull request by uncleGen**
## What changes were proposed in this pull request?

After this PR, we can create and register Hive UDFs to accept UDT type, like `VectorUDT` and `MatrixUDT`. These UDTs are widely used in Spark machine learning. 

## How was this patch tested?

add new ut

### CODE REVIEW

**Comment by uncleGen** [comment]
retest this please.

**Comment by uncleGen** [comment]
@cloud-fan Could you please help to review?

**Comment by maropu** [comment]
How do you handle these type data in Hive UDFs?

**Comment by uncleGen** [comment]
@maropu The internal sql type of `VectorUDT` and `MatrixUDT` is `StructType`. We can use `StructObjectInspector` in Hive UDFs to handle there type data. There is a demo: https://github.com/aliyun/aliyun-emapreduce-sdk/blob/master-2.x/emr-sql/src/main/scala/org/apache/spark/sql/aliyun/udfs/ml/LogisticRegressionUDF.scala#L86

**Comment by maropu** [comment]
But, the UDF works in Spark and not in Hive, right? I'm not sure that this is a right approach...

**Comment by uncleGen** [comment]
@maropu I dont test the UDF in Hive, but I think it will work as it is a Hive UDF definition.

**Comment by cloud-fan** [comment]
Can we add an end-to-end test to show the value of this patch?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveInspectorSuite.scala [hunk 1]
```diff
@@ -214,6 +214,15 @@ class HiveInspectorSuite extends SparkFunSuite with HiveInspectors {
     })
   }
 
+  test("wrap / unwrap UDT Type") {
+    case object TestUserClassUDT extends TestUserClassUDT
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveInspectorSuite.scala]: Why do we need this extend this?

File: mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala [hunk 2]
```diff
@@ -44,4 +45,39 @@ class VectorUDTSuite extends SparkFunSuite {
...
+      df.createOrReplaceTempView("src")
+
+      // `Logistic_Regression` accepts features (with Vector type), and returns the
+      // prediction value. To simplify the UDF implementation, the `Logistic_Regression`
+      // will return 0.95d directly.
+      sql(
+        s"""
+           |CREATE FUNCTION Logistic_Regression
+           |AS 'org.apache.spark.sql.hive.LogisticRegressionUDF'
+           |USING JAR '${TestHive.getHiveFile("TestLogRegUDF.jar").toURI}'
```
Comment (HyukjinKwon) [inline mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala]: What's the code in `TestLogRegUDF.jar`?

File: mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala [hunk 3]
```diff
@@ -44,4 +45,39 @@ class VectorUDTSuite extends SparkFunSuite {
     assert(dataType.asInstanceOf[StructType].fields.map(_.dataType)
       === Seq(new VectorUDT, DoubleType))
   }
+
+  test("SPARK-28158 Hive UDFs supports UDT type") {
```
Comment (cloud-fan) [inline mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala]: We can put this test in the sql/hive module, and create a UDT for testing.

Comment (uncleGen) [reply mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala]: ok

**Comment by HyukjinKwon** [comment]
ping @uncleGen

**Comment by uncleGen** [comment]
let me continue this pr

File: mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala (hunk 2 above)
Comment (uncleGen) [reply mllib/src/test/scala/org/apache/spark/ml/linalg/VectorUDTSuite.scala]: remove this jar

File: mllib/pom.xml [hunk 4]
```diff
@@ -74,6 +74,19 @@
       <type>test-jar</type>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (uncleGen) [reply mllib/pom.xml]: remove it

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveInspectorSuite.scala (hunk 1 above)
Comment (uncleGen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveInspectorSuite.scala]: fix

**Comment by uncleGen** [comment]
ping @HyukjinKwon @cloud-fan

**Comment by uncleGen** [comment]
ping @HyukjinKwon @cloud-fan

**Review by HyukjinKwon — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.spark.sql.hive.test.TestHive
+import org.apache.spark.sql.test.{ExamplePoint, ExamplePointUDT}
+import org.apache.spark.sql.types._
+
+class HiveUserDefinedTypeSuite extends QueryTest {
+  private[this] val functionClass = classOf[org.apache.spark.sql.hive.TestUDF].getCanonicalName
+
+  test("Support UDT in Hive UDF") {
+    val functionName = "get_point_x"
+    val sql = spark.sql _
```
[the remark this answers is not in the document]
Comment (uncleGen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: get it

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+      val input = inputGenerator.apply().asInstanceOf[Row]
+      val df = spark.createDataFrame(Array(input).toList.asJava, schema)
+      df.createOrReplaceTempView("src")
+      spark.sql(s"CREATE FUNCTION $functionName AS '$functionClass'")
+
+      checkAnswer(
+        spark.sql(s"SELECT $functionName(point) FROM src"),
+        Row(input.getAs[ExamplePoint](0).x))
+    } catch {
+      case cause: Throwable => throw cause
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: why do we need this line?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDF
+import org.apache.hadoop.hive.serde2.objectinspector.{ObjectInspector, StandardListObjectInspector}
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory
+
+import org.apache.spark.sql.{QueryTest, RandomDataGenerator, Row, SparkSession}
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.hive.test.TestHive
+import org.apache.spark.sql.test.{ExamplePoint, ExamplePointUDT}
+import org.apache.spark.sql.types._
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: nit: -> `import org.apache.spark.sql.types.StructType`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  override def getDisplayString(children: Array[String]): String = "get_point_x"
+
+  override def initialize(arguments: Array[ObjectInspector]): ObjectInspector = {
+    data = arguments(0).asInstanceOf[StandardListObjectInspector]
+    PrimitiveObjectInspectorFactory.javaDoubleObjectInspector
+  }
+
+  override def evaluate(arguments: Array[GenericUDF.DeferredObject]): AnyRef = {
+    val point = data.getList(arguments(0).get())
+    new lang.Double(point.get(0).asInstanceOf[Double])
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: nit: I would just do `java.lang.Double`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.hadoop.hive.serde2.objectinspector.{ObjectInspector, StandardListObjectInspector}
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory
+
+import org.apache.spark.sql.{QueryTest, RandomDataGenerator, Row, SparkSession}
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.hive.test.TestHive
+import org.apache.spark.sql.test.{ExamplePoint, ExamplePointUDT}
+import org.apache.spark.sql.types._
+
+class HiveUserDefinedTypeSuite extends QueryTest {
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: nit: why don't we use `with TestHiveSingleton`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        !spark.sessionState.catalog.functionExists(FunctionIdentifier(functionName)),
+        s"Function $functionName should have been dropped. But, it still exists.")
+    }
+  }
+
+  override protected def spark: SparkSession = TestHive.sparkSession
+}
+
+class TestUDF extends GenericUDF {
+  var data: StandardListObjectInspector = _
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: nit: private

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory
+
+import org.apache.spark.sql.{QueryTest, RandomDataGenerator, Row, SparkSession}
+import org.apache.spark.sql.catalyst.FunctionIdentifier
+import org.apache.spark.sql.hive.test.TestHive
+import org.apache.spark.sql.test.{ExamplePoint, ExamplePointUDT}
+import org.apache.spark.sql.types._
+
+class HiveUserDefinedTypeSuite extends QueryTest {
+  private[this] val functionClass = classOf[org.apache.spark.sql.hive.TestUDF].getCanonicalName
```
Comment (HyukjinKwon) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: nit: `private val`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala (hunk 6 above)
Comment (uncleGen) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveUserDefinedTypeSuite.scala]: removed

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HeartSaVioR** [comment]
I've submitted a follow-up PR as new test is high likely failing in recent CI builds. #26287

**Pull request closed without merging**